pinot-spi/src/main/java/org/apache/pinot/spi/env/PinotConfiguration.java

```diff
@@ -0,0 +1,250 @@
+  }
+
+  public String getProperty(String name, String defaultValue) {
+    return getRawProperty(name, defaultValue).toString();
+  }
+
+  private <T> T getProperty(String name, T defaultValue, Class<T> returnType) {
+    String relaxedPropertyName = relaxPropertyName(name);
+    if (!configuration.containsKey(relaxedPropertyName)) {
+      return defaultValue;
+    }
+
+    return PropertyConverter.convert(getRawProperty(name, defaultValue), returnType);
+  }
+
+  public boolean getProperty(String name, boolean defaultValue) {
+    return getProperty(name, defaultValue, Boolean.class);
+  }
+
+  public int getProperty(String name, int defaultValue) {
+    return getProperty(name, defaultValue, Integer.class);
+  }
+
+  public long getProperty(String name, long defaultValue) {
+    return getProperty(name, defaultValue, Long.class);
+  }
+
+  public double getProperty(String name, double defaultValue) {
+    return getProperty(name, defaultValue, Double.class);
+  }
+
+  private static Configuration loadProperties(String configPath) {
+    try {
+      PropertiesConfiguration propertiesConfiguration = new PropertiesConfiguration();
+
+      propertiesConfiguration.setIOFactory(new ConfigFilePropertyReaderFactory());
+      if (configPath.startsWith("classpath:")) {
+        propertiesConfiguration
+            .load(PinotConfiguration.class.getResourceAsStream(configPath.substring("classpath:".length())));
+      } else {
+        propertiesConfiguration.load(configPath);
+      }
+
+      return propertiesConfiguration;
+    } catch (ConfigurationException e) {
+      throw new RuntimeException("Could not read properties from " + configPath, e);
+    }
+  }
+
+  private static Map<String, Object> relaxConfigurationKeys(Configuration configuration) {
+    return CommonsConfigurationUtils.getKeysStream(configuration)
+
+        .collect(Collectors.toMap(PinotConfiguration::relaxPropertyName, key -> getProperty(key, configuration)));
+  }
+
+  private static Map<String, String> relaxEnvironmentVariables(Map<String, String> environmentVariables) {
+    return environmentVariables.entrySet().stream().filter(entry -> entry.getKey().startsWith("PINOT_"))
+        .collect(Collectors.toMap(PinotConfiguration::relaxEnvVarName, Entry::getValue));
+  }
+
+  private static String relaxEnvVarName(Entry<String, String> envVarEntry) {
+    return envVarEntry.getKey().substring(6).replace("_", ".").toLowerCase();
+  }
+
+  private static Map<String, Object> relaxProperties(Map<String, Object> properties) {
+    return properties.entrySet().stream()
+        .collect(Collectors.toMap(PinotConfiguration::relaxPropertyName, Entry::getValue));
+  }
+
+  private static String relaxPropertyName(Entry<String, Object> propertyEntry) {
+    return relaxPropertyName(propertyEntry.getKey());
+  }
+
+  private static String relaxPropertyName(String propertyName) {
+    return propertyName.replace("-", "").toLowerCase();
```

Comment:
I saw the property names would be converted to lower cases, that means e.g. `controller.upload.onlineToOfflineTimeout` will become `controller.upload.onlinetoofflinetimeout`. Is that intentional?

Reply:
This is an intentional behaviour. To support properties provided through environment variables, the relaxed binding mechanism sanitize all properties inserted and extracted from `PinotConfiguration`. There is no breaking change since a value configured with `controller.upload.onlineToOfflineTimeout` will be resolvable with the any of the following configs key: 

* `controller.upload.onlineToOfflineTimeout`
* `controller.upload.onlinetoofflinetimeout`
* `controller.upload.online-to-offline-timeout`
* `PINOT_CONTROLLER_UPLOAD_ONLINETOOFFLINETIMEOUT`